httpx/_models.py

```diff
@@ -515,8 +515,13 @@ def __setitem__(self, key: str, value: str) -> None:
         Set the header `key` to `value`, removing any duplicate entries.
         Retains insertion order.
         """
-        set_key = key.lower().encode(self.encoding)
-        set_value = value.encode(self.encoding)
+        try:
+            set_key = key.lower().encode(self.encoding)
+            set_value = value.encode(self.encoding)
+        except UnicodeEncodeError:
+            self.encoding = "utf8"
+            set_key = key.lower().encode(self.encoding)
+            set_value = value.encode(self.encoding)
```

Comment:
Actually I think we need a slightly different approach on this, since eg. we don't want to just change `self.encoding` if it has been explicitly set.

How's something like this?...

```python
encoding = self._encoding or "utf-8"
set_key = key.lower().encode(encoding)
set_value = value.encode(encoding)
```

Reply:
Done :-)